pkg/deploy/predeploy.go

```diff
@@ -439,25 +463,91 @@ func (d *deployer) createSecret(ctx context.Context, kv keyvault.Manager, secret
 	})
 }
 
-func (d *deployer) ensureSecretKey(ctx context.Context, kv keyvault.Manager, secretName string) error {
+func (d *deployer) ensureSecretKey(ctx context.Context, kv keyvault.Manager, secretName string) (isNew bool, err error) {
 	existingSecrets, err := kv.GetSecrets(ctx)
 	if err != nil {
-		return err
+		return false, err
 	}
 
 	for _, secret := range existingSecrets {
 		if filepath.Base(*secret.ID) == secretName {
-			return nil
+			return false, nil
 		}
 	}
 
 	key, err := rsa.GenerateKey(rand.Reader, 2048)
 	if err != nil {
-		return err
+		return false, err
 	}
 
 	d.log.Infof("setting %s", secretName)
-	return kv.SetSecret(ctx, secretName, azkeyvault.SecretSetParameters{
+	return true, kv.SetSecret(ctx, secretName, azkeyvault.SecretSetParameters{
 		Value: to.StringPtr(base64.StdEncoding.EncodeToString(x509.MarshalPKCS1PrivateKey(key))),
 	})
 }
+
+func (d *deployer) restartOldScalesets(ctx context.Context, resourceGroupName string) error {
+	d.log.Print("restarting old scalesets")
+	scalesets, err := d.vmss.List(ctx, resourceGroupName)
+	if err != nil {
+		return err
+	}
+
+	for _, vmss := range scalesets {
+		err = d.restartOldScaleset(ctx, *vmss.Name, resourceGroupName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *deployer) restartOldScaleset(ctx context.Context, vmssName string, resourceGroupName string) error {
+	var restartScript string
+	var waitForReadiness func(ctx context.Context, vmssName string) error
+	switch {
+	case strings.HasPrefix(vmssName, gatewayVMSSPrefix):
+		restartScript = gatewayRestartScript
+		waitForReadiness = d.gatewayWaitForReadiness
+	case strings.HasPrefix(vmssName, rpVMSSPrefix):
+		restartScript = rpRestartScript
+		waitForReadiness = d.rpWaitForReadiness
+	default:
+		return nil
+	}
```

Comment:
I am not entirely convinced we need this wait for readiness. I checked the documentation of the command we are calling https://learn.microsoft.com/en-us/rest/api/compute/virtual-machine-scale-set-vms/run-command?tabs=HTTP#runcommandinput and it is supposed to by simple call that run command in the running VM.

Moreover we are not rebooting the VMs, so they will not change they readiness probe. They will always be ready. And this is the state we are checking in the readiness checks:
```
instanceUnhealthy := r.VMHealth != nil && r.VMHealth.Status != nil && r.VMHealth.Status.Code != nil && *r.VMHealth.Status.Code != "HealthState/healthy"
```

Up to my best understanding to the Azure API. We should be fine by simply waiting for the command output, check that and done.

Reply:
Thanks @petrkotas for providing the api links and insight. I am not sure if the readiness probe will always be the same, according to [load-balancer custom probe doc](https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-custom-probe-overview#probe-interval), if the probe fails it will mark the backend-pool instance unhealthy. And as there's a LB health probe named rp-probe that checks the rp connection, there can be a chance the `run-command` returns successfully, but the rp connection takes time to become healthy.
Am I thinking correctly?

Reply:
@petrkotas we define a custom probe for RP here https://github.com/Azure/ARO-RP/blob/master/pkg/deploy/assets/rp-production.json#L283-L289 that is used for readiness, so the state of the probe will change depending on the state of the RP service itself.